test/document.js

```diff
@@ -565,7 +565,31 @@ describe('serialize document', function() {
 
     assert.throws(() => {
       firestore.doc('collectionId/documentId').update(obj);
-    }, new RegExp('Argument "dataOrField" is not a valid Document. Input ' + 'object is deeper than 20 levels or contains a cycle.'));
+    }, new RegExp('Argument "dataOrField" is not a valid Document. Input object is deeper than 20 levels or contains a cycle.'));
+  });
+
+  it('is able to write a document reference with cycles', function() {
+    firestore.api.Firestore._commit = function(request, options, callback) {
+      requestEquals(
+        request,
+        set(
+          document('ref', {
+            referenceValue:
+              'projects/test-project/databases/(default)/documents/collectionId/documentId',
+            valueType: 'referenceValue',
+          })
+        )
+      );
+
+      callback(null, defaultWriteResult);
+    };
+
+    // The Firestore Admin SDK adds a cyclic reference to the 'Firestore' type.
+    // We emulate this behavior in this test to verify that we can properly
+    // serialize types that contain references to a cyclic Firestore type.
```

Comment:
"types that contain references to a cyclic Firestore type" is broader than we actually should support.  I think it should just be "DocumentReference instances, even if they have cyclic references (we shouldn't try to validate them beyond the instanceof check)."

Reply:
Done